pkg/component/controller/metrics.go

```diff
@@ -0,0 +1,359 @@
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	log := logrus.WithFields(logrus.Fields{"component": "Metrics"})
+
+	restClient, err := clientCF.GetRESTClient()
+	if err != nil {
+		return nil, fmt.Errorf("error getting REST client for metrics: %w", err)
+	}
+
+	return &Metrics{
+		hostname:   hostname,
+		log:        log,
+		K0sVars:    k0sVars,
+		saver:      saver,
+		restClient: restClient,
+	}, nil
+}
+
+// Init does nothing
+func (m *Metrics) Init(_ context.Context) error {
+	tw := templatewriter.TemplateWriter{
+		Name:     "pushgateway-with-ttl",
+		Template: pushGatewayTemplate,
+		Data: map[string]string{
+			"Namespace": namespace,
+			"Name":      pushGatewayName,
+		},
+	}
+	output := bytes.NewBuffer([]byte{})
+	err := tw.WriteToBuffer(output)
+	if err != nil {
+		return err
+	}
+	err = m.saver.Save("pushgateway.yaml", output.Bytes())
+	if err != nil {
+		return err
+	}
+
+	var j *job
+	j, err = m.newJob("kube-scheduler", "https://localhost:10259/metrics")
+	if err != nil {
+		return err
+	}
+	m.jobs = append(m.jobs, j)
+
+	j, err = m.newJob("kube-controller-manager", "https://localhost:10257/metrics")
+	if err != nil {
+		return err
+	}
+	m.jobs = append(m.jobs, j)
+
+	return nil
+}
+
+// Run runs the metric server reconciler
+func (m *Metrics) Run(ctx context.Context) error {
+	ctx, m.tickerDone = context.WithCancel(ctx)
+
+	for _, j := range m.jobs {
+		go j.Run(ctx)
+	}
+
+	return nil
+}
+
+// Stop stops the reconciler
+func (m *Metrics) Stop() error {
+	if m.tickerDone != nil {
+		m.tickerDone()
+	}
+	return nil
+}
+
+// Reconcile detects changes in configuration and applies them to the component
+func (m *Metrics) Reconcile(_ context.Context, clusterConfig *v1beta1.ClusterConfig) error {
+	m.log.Debug("reconcile method called for: Metrics")
+
+	// We just store the last known config
+	for _, j := range m.jobs {
+		j.clusterConfig = clusterConfig
+	}
+	m.clusterConfig = clusterConfig
+	return nil
+}
+
+// Healthy is the health-check interface
+func (m *Metrics) Healthy() error { return nil }
+
+type job struct {
+	scrapeURL     string
+	name          string
+	hostname      string
+	clusterConfig *v1beta1.ClusterConfig
+	log           *logrus.Entry
+	scrapeClient  *http.Client
+	restClient    rest.Interface
+}
+
+func (m *Metrics) newJob(name, scrapeURL string) (*job, error) {
+	certFile := path.Join(m.K0sVars.CertRootDir, "admin.crt")
+	keyFile := path.Join(m.K0sVars.CertRootDir, "admin.key")
+
+	httpClient, err := getClient(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return &job{
+		scrapeURL:    scrapeURL,
+		name:         name,
+		hostname:     m.hostname,
+		log:          m.log,
+		scrapeClient: httpClient,
+		restClient:   m.restClient,
+	}, nil
+}
+
+func (j *job) Run(ctx context.Context) {
+	j.log.Debugf("Running %s job", j.name)
+
+	t := time.NewTicker(time.Second * 30)
```

Comment:
Is 30sec bit too often?

Reply:
Yeah, maybe. I put 30s here because it's the default prometheus scrape_interval. What value are you suggesting?

Reply:
oh, somehow I remembered prom scrape default was 60s. I think 30s is fine